### PR TITLE
Implement cache mechanism for tracing data

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description
This PR is a Proof of Concept about using a cache mechanism for tracing API.
The main idea is to pre-generate block traces on the go, while the chain progresses.
## Fixed Issue(s)

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.function.Supplier;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.ehcache.Cache;
+import org.ehcache.CacheManager;
+
+public class TracingCacheManager {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int CACHE_SIZE = 1000;
```
Comment (atoulme) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Can this be made configurable?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.ehcache.Cache;
+import org.ehcache.CacheManager;
+
+public class TracingCacheManager {
```
Comment (atoulme) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Can tests target this class directly as unit test?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            blockTrace ->
+                blockTraceCache.put(blockContext.getBlockHeader().getNumber(), blockTrace));
+  }
+
+  public Optional<BlockTrace> blockTraceAt(final long blockNumber) {
+    LOG.info("Looking for cache at block number: {}", blockNumber);
+    LOG.info(
+        "Cached data present for block {}: {}",
+        blockNumber,
+        blockTraceCache.containsKey(blockNumber));
```
Comment (atoulme) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: This statement executes. Please consider wrapping with `if (LOG.isInfoEnabled())`

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            .build(true)) {
+      blockTraceCache = cacheManager.getCache(CACHE_NAME, Long.class, BlockTrace.class);
+    }
+    besuContext
+        .getService(BesuEvents.class)
+        .ifPresent(besuEvents -> besuEvents.addBlockPropagatedListener(this::onNewBlock));
+  }
+
+  public void onNewBlock(final PropagatedBlockContext blockContext) {
+    LOG.info("New block detected, starting to cache tracing data.");
```
Comment (atoulme) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Consider moving down to debug level.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Totally, that would be the idea

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Sure.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java (hunk 3 above)
Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Yeah this is just in the context of the POC.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java (hunk 4 above)
Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/cache/TracingCacheManager.java]: Yeah this is just in the context of the POC.

**Pull request closed without merging**